test/gtest/wat-parser.cpp

```diff
@@ -0,0 +1,337 @@
+#include "gtest/gtest.h"
+
+#include "wat-parser-internal.h"
+
+using namespace wasm::WATParser;
+
+TEST(ParserTest, LexWhitespace) {
+  Token one{"1"sv, IntTok{1, false}};
+  Token two{"2"sv, IntTok{2, false}};
+  Token three{"3"sv, IntTok{3, false}};
+  Token four{"4"sv, IntTok{4, false}};
+  Token five{"5"sv, IntTok{5, false}};
+
+  Lexer lexer(" 1\t2\n3\r4 \n\n\t 5 "sv);
+
+  auto it = lexer.begin();
+  ASSERT_NE(it, lexer.end());
+  Token t1 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t2 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t3 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t4 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t5 = *it++;
+  EXPECT_EQ(it, lexer.end());
+
+  EXPECT_EQ(t1, one);
+  EXPECT_EQ(t2, two);
+  EXPECT_EQ(t3, three);
+  EXPECT_EQ(t4, four);
+  EXPECT_EQ(t5, five);
+
+  EXPECT_EQ(lexer.position(t1), (TextPos{1, 1}));
+  EXPECT_EQ(lexer.position(t2), (TextPos{1, 3}));
+  EXPECT_EQ(lexer.position(t3), (TextPos{2, 0}));
+  EXPECT_EQ(lexer.position(t4), (TextPos{2, 2}));
+  EXPECT_EQ(lexer.position(t5), (TextPos{4, 2}));
+}
+
+TEST(ParserTest, LexLineComment) {
+  Token one{"1"sv, IntTok{1, false}};
+  Token six{"6"sv, IntTok{6, false}};
+
+  Lexer lexer("1;; whee! 2 3\t4\r5\n6"sv);
+
+  auto it = lexer.begin();
+  Token t1 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t2 = *it++;
+  EXPECT_EQ(it, lexer.end());
+
+  EXPECT_EQ(t1, one);
+  EXPECT_EQ(t2, six);
+
+  EXPECT_EQ(lexer.position(t1), (TextPos{1, 0}));
+  EXPECT_EQ(lexer.position(t2), (TextPos{2, 0}));
+}
+
+TEST(ParserTest, LexBlockComment) {
+  Token one{"1"sv, IntTok{1, false}};
+  Token six{"6"sv, IntTok{6, false}};
+
+  Lexer lexer("1(; whoo! 2\n (; \n3\n ;) 4 (;) 5 ;) \n;)6"sv);
+
+  auto it = lexer.begin();
+  Token t1 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t2 = *it++;
+  EXPECT_EQ(it, lexer.end());
+
+  EXPECT_EQ(t1, one);
+  EXPECT_EQ(t2, six);
+
+  EXPECT_EQ(lexer.position(t1), (TextPos{1, 0}));
+  EXPECT_EQ(lexer.position(t2), (TextPos{5, 2}));
+}
+
+TEST(ParserTest, LexParens) {
+  Token left{"("sv, LParenTok{}};
+  Token right{")"sv, RParenTok{}};
+
+  Lexer lexer("(())"sv);
+
+  auto it = lexer.begin();
+  ASSERT_NE(it, lexer.end());
+  Token t1 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t2 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t3 = *it++;
+  ASSERT_NE(it, lexer.end());
+  Token t4 = *it++;
+  EXPECT_EQ(it, lexer.end());
+
+  EXPECT_EQ(t1, left);
+  EXPECT_EQ(t2, left);
+  EXPECT_EQ(t3, right);
+  EXPECT_EQ(t4, right);
+}
+
+TEST(ParserTest, LexInt) {
+  {
+    Lexer lexer("0"sv);
+    ASSERT_NE(lexer, lexer.end());
+    Token expected{"0"sv, IntTok{0, false}};
+    EXPECT_EQ(*lexer, expected);
+  }
+  {
+    Lexer lexer("+0"sv);
+    ASSERT_NE(lexer, lexer.end());
+    Token expected{"+0"sv, IntTok{0, true}};
+    EXPECT_EQ(*lexer, expected);
+  }
+  {
+    Lexer lexer("-0"sv);
+    ASSERT_NE(lexer, lexer.end());
+    Token expected{"-0"sv, IntTok{0, true}};
```

Comment:
We need to detect that negative zero is different than positive zero, so it it's a little odd that the `IntTok` is the same in the last two cases. I see @aheejin asked the same thing in the source code above.

Reply:
We don't need to detect any difference between `+0` and `-0`, but we do need to detect the difference between those and plain `0`. The signed versions are valid only where the grammar expects `sN` or `iN` for some `N`, but not where it expects `uN`.

Reply:
Are you sure? We have spec tests that check for the behavior of negative zero on various math ops, which can differ from "normal" zero. Those depend on being able to parse it. Or maybe we mean something different by "detect"?

Reply:
Maybe you're thinking about floating point zeroes? Those are definitely different and would have spec tests checking their behavior. But negative and positive integer zero are just two ways of writing the same signed integer value.

Reply:
Oh, ok, yes, I was thinking of floats...